docs/RFCS/replicate_queue_remove_replicas.md

```diff
@@ -0,0 +1,166 @@
+`allocateTargetInternal()`, `RebalanceTarget()` and `ShouldRebalance()` which
+determine how a given store's used capacity relates to the mean used capacity
+for the cluster.
+
+This code will be refactored into three functions:
++ `bestStore()`, which will return the "best" store in a slice of stores according to the
+  capacity criteria.
++ `worstStore()`, which returns the "worst" store in a slice.
++ `evaluateStore()` function, which will simply determine if a store is "good"
+  or "bad" according to the overall state of the cluster.
+
+The functionality within allocator will be refactored to use these functions.
+
+#### replicateQueue.process
+Process will now use the following logic for a replica which requires a replication change:
+
+1. If the range needs additional replicas:
+  + The queue will use the allocator's `AllocateTarget()` method to select a target node.
+  + If a target node is found, the queue will attempt to add a replica to it.
+2. If the range has too many replicas:
+  + The queue will grade the existing replicas using `allocator.worstStore()`.
+  + The worst replica will be removed.
+3. In either case, the queued replica will be re-queued to check for additional
+   changes, unless it was the replica that was removed.
+
+## Distribution Concerns
+The replication queue process can occur on any replica of a range; there is no
+requirement that it runs only on the leader or any other specific replica.
+Therefore, it is possible that multiple replicas for the same range could be
+placed in the queue at the same time.
+
+At the same time, many of the stats used to evaluate the allocator's methods
+(e.g. `evaluateStore()`) are derived from gossip statistics, which may be
+inconsistent across stores.  Therefore, the queues on different stores could
+come to different conclusions on which replica should be added or removed. If
+the transactions are not properly coordinated, this could result in double adds
+or double removes.
+
+The existing ChangeReplicas does have some protection against races, but it is
+not adequate in our case. 
+
+This problem is specially captured in issue [#2152](https://github.com/cockroachdb/cockroach/issues/2152), which includes a solution.
+That issue is a prequisite for this RFC.
+
+# Drawbacks
+
+# Alternatives
+
+### Two queues
+A simple alternative to the design would split the replicate queue into two
+queues, an 'up-replicate' and 'down-replicate' queue. 
+
+However, this does not make the function of the queue significantly clearer; the
+goal of "right-sizing" the replication factor for a range is very clear, and
+dividing it further does not add clarity.
+
+### One queue for all possible changes
+Another queue-based alternative (which represents further work) would be to combine the
+rebalance, repair and replicate queues into a single queue. This queue would be
+able to make more complicated decisions about the replication state of a range.
```

Comment:
should add here that an upside would be that you'd be able to avoid conflicts between different actors (as in the comments above) more easily.